qp_klp/Step.py

```diff
@@ -369,32 +386,140 @@ def generate_sifs(self, from_qiita):
+        self.tube_id_map = tids_by_qiita_id
+        self.samples_in_qiita = sample_names_by_qiita_id
+
+        # not needed, but useful in displaying debugging info
+        return tids_by_qiita_id
+
+    def compare_samples_against_qiita(self):
+        projects = self.pipeline.get_project_info(short_names=True)
+
+        results = []
+        for project in projects:
+            project_name = project['project_name']
+            qiita_id = str(project['qiita_id'])
+
+            # get list of samples as presented by the sample-sheet or mapping
+            # file and confirm that they are all registered in Qiita.
+            samples = set(self.pipeline.get_sample_names())
+
+            # strip any leading zeroes from the sample-ids. Note that
+            # if a sample-id has more than one leading zero, all of
+            # them will be removed.
+            samples = {x.lstrip('0') for x in samples}
+
+            # just get a list of the tube-ids themselves, not what they map
+            # to.
+            if qiita_id in self.tube_id_map:
+                # if map is not empty
+                tids = [self.tube_id_map[qiita_id][x] for x in
+                        self.tube_id_map[qiita_id]]
+                not_in_qiita = samples - set(tids)
+                examples = tids[:5]
+                used_tids = True
+            else:
+                # assume project is in samples_in_qiita
+                not_in_qiita = samples - set(self.samples_in_qiita)
+                examples = list(samples)[:5]
+                used_tids = False
+
+            # convert to strings before returning
+            examples = [str(x) for x in examples]
+
+            if not_in_qiita:
+                # if there are actual differences
+                results.append({'samples_not_in_qiita': not_in_qiita,
+                                'examples_in_qiita': examples,
+                                'project_name': project_name,
+                                'tids': used_tids})
+        return results
+
+    @classmethod
+    def _overwrite_prep_file(cls, prep_file_path, qiita_id, tube_id_map):
+        # passing tube_id_map as a parameter allows for easier testing.
+        df = pd.read_csv(prep_file_path, sep='\t', dtype=str, index_col=False)
+        # save copy of sample_name column as 'old_sample_name'
+        df['old_sample_name'] = df['sample_name']
+        for i in df.index:
+            sample_name = df.at[i, "sample_name"]
+            # blanks do not get their names swapped.
+            if sample_name.startswith('BLANK'):
+                continue
+
+            # remove leading zeroes if they exist to match Qiita results.
+            sample_name = sample_name.lstrip('0')
+            if sample_name in tube_id_map[qiita_id]:
+                df.at[i, "sample_name"] = tube_id_map[qiita_id][sample_name]
+
+        df.to_csv(prep_file_path, index=False, sep="\t")
+
+    def _overwrite_prep_files(self, prep_file_paths):
```

Comment:
The classmethod _overwrite_prep_file and this one are really similar, maybe select something less similar to avoid confusion? Also, it looks like the only place where _overwrite_prep_file is called is within _overwrite_prep_files; do we really need that split?

Reply:
I can change the names to be more unique but I'd like to keep it as two separate functions. I'll be writing an extra test for the static method.